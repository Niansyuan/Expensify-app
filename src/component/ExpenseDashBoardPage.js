import React from 'react'

const ExpenseDashBoardPage=()=>(
    <div>
        <p>This is from dashBoard component</p>
    </div>
)

export default ExpenseDashBoardPage