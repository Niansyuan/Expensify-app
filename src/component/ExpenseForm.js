import "react-dates/initialize";
import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
    state={
        description:'',
        note:'',
        amount:'',
        createAt:moment(),
        calendarFocused: false,
        error:''
    };
    handleDiscriptionChange=(e)=>{
        const description=e.target.value;
        this.setState(()=>({description}));
    }
    handleNoteChange=(e)=>{
        const note=e.target.value;
        this.setState(()=>({note}));
    }
    handleAmountChange=(e)=>{
        const amount=e.target.value;
        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //regular expression,詳看notion筆記
            this.setState(()=>({amount}))
        }
    }
    handleDateChange=(createAt)=>{
        this.setState(()=>({createAt}))
    }
    handleFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused:focused}))
    }
    handleSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please Provide Description and Amount'}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10),
                createAt:this.state.createAt.valueOf(),
                note:this.state.note
            })
        }
    }
    render(){
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.handleSubmit}>
                    <input 
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.handleDiscriptionChange}
                    />
                    <input
                    type="number"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.handleAmountChange}
                    />
                    <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.handleDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.handleFocusChange}
                    numberOfMonths={1} //一次顯示幾個月
                    isOutsideRange={()=>false} 
                    />
                    <textarea
                    placeholder="Add a note for your expense...(optional)"
                    value={this.state.note}
                    onChange={this.handleNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}