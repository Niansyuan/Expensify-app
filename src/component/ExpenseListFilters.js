import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters.js";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    handleDateChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    handleFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        };
        if (e.target.value === "amount") {
            this.props.sortByAmount()
        }
    }
    render() {
        return (
            <div className="contain-container">
                <div className="filter">
                    <div className="filter__contain">
                        <input
                            className="text-input"
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                            placeholder='Search expenses' />
                    </div>
                    <div className="filter__contain">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="filter__contain">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={"startDate"}
                            endDate={this.props.filters.endDate}
                            endDateId={"endDate"}
                            onDatesChange={this.handleDateChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.handleFocusChange}
                            numberOfMonths={1} //一次顯示幾個月
                            isOutsideRange={() => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);