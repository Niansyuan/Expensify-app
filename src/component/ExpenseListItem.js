import React from "react";
import {Link} from 'react-router-dom'

const ExpenseListItem=({id,description,amount,createAt})=>{
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>${amount}--{createAt} ms</p> 
        </div>
    )
}

export default ExpenseListItem;