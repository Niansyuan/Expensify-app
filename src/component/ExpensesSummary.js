import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/expense-total';
import numeral from 'numeral';


export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const expenseTotalFormat = numeral(expenseTotal).format('$0,0.00');
    return (
        <div>
            <h1>viewing {expenseCount} {expenseWord} totalling NT{expenseTotalFormat}</h1>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: state.expenses.length,
        expenseTotal: selectExpenseTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)