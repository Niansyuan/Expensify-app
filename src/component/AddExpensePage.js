import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import {addExpenses} from '../actions/expenses'

const AddExpensePage=(props)=>(
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense)=>{
                props.dispatch(addExpenses(expense))
                props.history.push('/')
            }}
        />
    </div>
)

export default connect()(AddExpensePage);