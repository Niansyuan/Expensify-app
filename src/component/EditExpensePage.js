import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import {editExpenses,removeExpenses} from '../actions/expenses'

const EditExpensePage=(props)=>{
    console.log(props)
    return (
        <div>
            <ExpenseForm 
                expense={props.expense}
                onSubmit={(expense)=>{
                    props.dispatch(editExpenses(props.expense.id,expense))
                    props.history.push('/')
                    console.log('update',expense)
                }}
            />
            <button 
                expense={props.expense}
                onClick={()=>{
                props.dispatch(removeExpenses({id:props.expense.id}))
                props.history.push('/')
            }}>Remove</button>
        </div>
    )
}

const mapStateToProps=(state,props)=>{
    return {
        expense: state.expenses.find((expense)=>expense.id === props.match.params.id)
    }
}

export default connect(mapStateToProps)(EditExpensePage);