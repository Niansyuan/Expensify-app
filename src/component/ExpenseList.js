import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpense from "../selectors/expenses";

const ExpenseList=(props)=>(
    <div>
        <h1>Expense List</h1>
        {props.expenses.map((expense)=>{
            return <ExpenseListItem key={expense.id} {...expense}/>
        })}
    </div>
)

//HOC -------
const mapStateToProps=(state)=>{
    return {
        expenses:selectExpense(state.expenses,state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList);
//HOC -------