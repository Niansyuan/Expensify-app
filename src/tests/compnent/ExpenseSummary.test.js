import React from 'react';
import {shallow} from 'enzyme';
import { ExpensesSummary } from '../../component/ExpensesSummary';
import '../../tests/setupTests'

test('should correctly render expensesSummary with one expense',()=>{
    const wrapper=shallow(<ExpensesSummary expenseCount={1} expenseTotal={350}/>)
    expect(wrapper).toMatchSnapshot();
})

test('should correctly render expensesSummary with multiple expenses',()=>{
    const wrapper=shallow(<ExpensesSummary expenseCount={3} expenseTotal={4700}/>)
    expect(wrapper).toMatchSnapshot();
})