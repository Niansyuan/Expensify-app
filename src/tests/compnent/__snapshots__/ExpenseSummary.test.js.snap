// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    viewing 
    3
     
    expenses
     totalling 
    4,700.0
  </h1>
</div>
`;

exports[`should correctly render expensesSummary with one expense 1`] = `
<div>
  <h1>
    viewing 
    1
     
    expense
     totalling 
    350.0
  </h1>
</div>
`;
